radixdlt-engine/src/main/java/com/radixdlt/constraintmachine/ConstraintMachine.java

```diff
@@ -220,233 +168,86 @@ public SubstateCursor shutdownAll(Class<? extends Particle> particleClass) {
+		var requiredLevel = authorization.permissionLevel();
+		context.verifyPermissionLevel(requiredLevel);
 		try {
-			if (validationState.permissionLevel != PermissionLevel.SYSTEM) {
-				methodProcedure.verifyAuthorization(authorizationParam, readable, validationState.signedBy);
+			if (context.permissionLevel() != PermissionLevel.SYSTEM) {
+				if (requiredLevel == PermissionLevel.USER) {
+					this.metering.onUserInstruction(procedure.key(), procedureParam, context);
+				}
+
+				authorization.authorizer().verify(readableAddrs, context);
 			}
-			reducerResult = methodProcedure.call(procedureParam, reducerState, readable);
+
+			return procedure.call(procedureParam, reducerState, readableAddrs, context).state();
 		} catch (AuthorizationException e) {
-			throw new ConstraintMachineException(CMErrorCode.AUTHORIZATION_ERROR, validationState, e);
+			throw new ConstraintMachineException(CMErrorCode.AUTHORIZATION_ERROR, e);
 		} catch (ProcedureException e) {
-			throw new ConstraintMachineException(CMErrorCode.PROCEDURE_ERROR, validationState, e);
+			throw new ConstraintMachineException(CMErrorCode.PROCEDURE_ERROR, e);
 		} catch (Exception e) {
-			throw new ConstraintMachineException(CMErrorCode.UNKNOWN_ERROR, validationState, e);
+			throw new ConstraintMachineException(CMErrorCode.UNKNOWN_ERROR, e);
 		}
-
-		reducerResult.ifCompleteElse(
-			() -> validationState.reducerState = null,
-			nextState -> validationState.reducerState = nextState
-		);
 	}
 
 	/**
 	 * Executes transition procedures and witness validators in a particle group and validates
 	 * that the particle group is well formed.
 	 */
-	void statefulVerify(
+	List<List<REStateUpdate>> statefulVerify(
+		ExecutionContext context,
 		CMValidationState validationState,
-		List<REInstruction> instructions,
-		List<List<REStateUpdate>> parsedInstructions
+		List<REInstruction> instructions
 	) throws ConstraintMachineException {
 		int instIndex = 0;
 		var expectEnd = false;
-
-		var parsed = new ArrayList<REStateUpdate>();
+		ReducerState reducerState = null;
+		var readableAddrs = validationState.immutableIndex();
+		var groupedStateUpdates = new ArrayList<List<REStateUpdate>>();
+		var stateUpdates = new ArrayList<REStateUpdate>();
 
 		for (REInstruction inst : instructions) {
-			validationState.curIndex = instIndex;
-			validationState.curInstruction = inst;
-
-			if (expectEnd && inst.getMicroOp() != REInstruction.REOp.END) {
-				throw new ConstraintMachineException(CMErrorCode.MISSING_PARTICLE_GROUP, validationState);
+			if (expectEnd && inst.getMicroOp() != REInstruction.REMicroOp.END) {
+				throw new ConstraintMachineException(CMErrorCode.MISSING_PARTICLE_GROUP);
 			}
 
-			if (inst.getMicroOp() == REInstruction.REOp.DOWNALL) {
+			if (inst.getMicroOp() == REInstruction.REMicroOp.SYSCALL) {
+				CallData callData = inst.getData();
+				var opSignature = OpSignature.ofMethod(inst.getMicroOp().getOp(), REAddr.ofSystem());
```

Comment:
Do you think it's worth to make the operand of `SYSCALL` generic, such as `address + bytes`? The fee taking system contract would parse a `u256` from the `bytes`.

Reply:
Yup, the `CallData` object is supposed to be the generic operand. Avoiding use of `byte[]` so that we can possibly avoid buffer copies.

Reply:
As discussed offline, we're not adding address ATM. Instead, we will reserve a few bytes in the calldata to distinguish different actions.